src/backend/dx12/src/device.rs

```diff
@@ -3407,13 +3409,133 @@ impl d::Device<B> for Device {
+            SampleDesc: dxgitype::DXGI_SAMPLE_DESC {
+                Count: 1,
+                Quality: 0,
+            },
+            Layout: d3d12::D3D12_TEXTURE_LAYOUT_ROW_MAJOR,
+            Flags: d3d12::D3D12_RESOURCE_FLAG_NONE,
+        };
+
+        let properties = d3d12::D3D12_HEAP_PROPERTIES {
+            Type: d3d12::D3D12_HEAP_TYPE_READBACK,
+            CPUPageProperty: d3d12::D3D12_CPU_PAGE_PROPERTY_UNKNOWN,
+            MemoryPoolPreference: d3d12::D3D12_MEMORY_POOL_UNKNOWN,
+            CreationNodeMask: 0,
+            VisibleNodeMask: 0,
+        };
+
+        let heap_desc = d3d12::D3D12_HEAP_DESC {
+            SizeInBytes: size,
+            Properties: properties,
+            Alignment: 0,
+            Flags: d3d12::D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS,
+        };
+
+        let mut heap = native::Heap::null();
+        assert_eq!(
+            self.raw
+                .clone()
+                .CreateHeap(&heap_desc, &d3d12::ID3D12Heap::uuidof(), heap.mut_void()),
+            winerror::S_OK
+        );
+
+        let mut temp_buffer = native::Resource::null();
+        assert_eq!(
+            winerror::S_OK,
+            self.raw.clone().CreatePlacedResource(
+                heap.as_mut_ptr(),
+                0,
+                &buffer_desc,
+                d3d12::D3D12_RESOURCE_STATE_COPY_DEST,
+                ptr::null(),
+                &d3d12::ID3D12Resource::uuidof(),
+                temp_buffer.mut_void(),
+            )
+        );
+
+        let list_type = native::CmdListType::Direct;
+        let (com_allocator, hr_alloc) = self.raw.create_command_allocator(list_type);
+        assert_eq!(
+            winerror::S_OK,
+            hr_alloc,
+            "error on command allocator creation: {:x}",
+            hr_alloc
+        );
+        let (com_list, hr_list) = self.raw.create_graphics_command_list(
+            list_type,
+            com_allocator,
+            native::PipelineState::null(),
+            0,
+        );
+        assert_eq!(
+            winerror::S_OK,
+            hr_list,
+            "error on command list creation: {:x}",
+            hr_list
+        );
+
+        let query_ty = match pool.ty {
+            query::Type::Occlusion => d3d12::D3D12_QUERY_TYPE_OCCLUSION,
+            query::Type::PipelineStatistics(_) => d3d12::D3D12_QUERY_TYPE_PIPELINE_STATISTICS,
+            query::Type::Timestamp => d3d12::D3D12_QUERY_TYPE_TIMESTAMP,
+        };
+        com_list.ResolveQueryData(
+            pool.raw.as_mut_ptr(),
+            query_ty,
+            queries.start,
+            num_queries,
+            temp_buffer.as_mut_ptr(),
+            0,
+        );
+        com_list.close();
+
+        self.queues[0]
+            .raw
+            .ExecuteCommandLists(1, &(com_list.as_mut_ptr() as *mut _));
+
+        if !flags.contains(query::ResultFlags::WAIT) {
+            warn!("Can't really not wait here")
+        }
+        self.queues[0].wait_idle_impl()?;
```

Comment:
This is an opportunity to exit before the resources are released below, does `d3d12-rs` have a "strong" `WeakPtr`?

Reply:
I think it does, but we aren't using it. You are correct though, I should make sure these are released.